### PR TITLE
Ticket #2323: Domain manager csv export

### LINKED ISSUE #2323: Multi-domain portfolio export to CSV file
### Issue description

Org administrators and multi-domain managers need to download ALL the domains and domain requests under their purview. This needs to result in a CSV file which may easily be imported into spreadsheet for analysis.

This data must be complete and current.

### Acceptance criteria

- [ ] `Org Admin `can download CSV output of all Organization common data, related data, and detailed records of all domains and domain requests in their organization.
- [ ] `Domain managers who are responsible for more than one domain` can download CSV output of all domains and domain requests for which they are a manager.

### Additional context

For any `Organization`, generate CSV report output with ALL relevant
All Org Common Data
List of Org Members
List of Org Domains
	All domain data
List of Org Domain Requests
	All domain request data

For a Domain Manager who handles multiple domains:
List of Domains
	All domain data, including other domain managers
List of Domain Requests

### Links to other issues

Based on:
- #2127

### PR DESCRIPTION
**Pull request by zandercymatics**
## Ticket

Resolves #2323

## Changes

- Adds a csv export button to the Domain table in home
- Small change to allow our base csv_export classes to be more flexible
- Tests

## Context for reviewers
**Note:** The redesign for the portfolio page has not been merged yet. The underlying design of the export button will not change, but the design of the portfolio page will - hence why we still see that.

This PR features two variations of the csv export button:
1. The csv export button when the `organization_feature` flag is on, and the given user has a portfolio
2. The csv button when the `organization_feature` flag is off and the given user doesn't have a portfolio

## Setup
1. In /admin, go to UserGroups and add yourself to a couple of domains
2. Navigate to /home with the organization_feature off
3. Note the csv export button at the bottom of the table. Click it, and ensure that the report matches whats on the table.
4. In /admin, click `Waffle Flags` and click the `organization_feature` record. Set everyone => yes
5. Create a portfolio, set yourself as the creator, then add a portfolio to a domain you manage
6. Navigate to /home
7. Note the csv export button by the search bar. Click it, and ensure that the report matches whats on the table.

#### Satisfied acceptance criteria and met development standards

- [ ] Met the acceptance criteria, or will meet them in a subsequent PR
- [ ] Created/modified automated tests
- [ ] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [ ] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [ ] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

### CODE REVIEW

File: src/registrar/templates/includes/domains_table.html [hunk 1]
```diff
@@ -167,6 +178,18 @@ <h2>Status</h2>
     <div class="domains__no-search-results display-none">
       <p>No results found</p>
     </div>
+
+    {% if not portfolio %}
+    <div class="grid-row">
```
Comment (rachidatecs) [inline src/registrar/templates/includes/domains_table.html:183]: Put your ifs around the div and close div markups, don't repeat the button markup. Haven't looked at this in a browser so hold off a bit.

File: src/registrar/assets/sass/_theme/_base.scss [hunk 2]
```diff
@@ -223,3 +223,9 @@ abbr[title] {
 .padding--8-8-9 {
   padding: 8px 8px 9px !important;
 }
+
+.usa-icon.usa-icon--big {
```
Comment (rachidatecs) [inline src/registrar/assets/sass/_theme/_base.scss:227]: You probably only need .usa-icon--big as a selector, you probably want to put this in the _buttons file

File: src/registrar/assets/sass/_theme/_buttons.scss [hunk 3]
```diff
@@ -176,4 +176,3 @@ a.usa-button--unstyled:visited {
     box-shadow: none;
   }
 }
-
```
Comment (rachidatecs) [inline src/registrar/assets/sass/_theme/_buttons.scss:179]: end line

**Review by rachidatecs — COMMENTED** [review]
Haven't tested this yet, just code comments

**Comment by gabydisarli** [comment]
@zandercymatics I'll be doing design review

**Comment by gabydisarli** [comment]
@zandercymatics a few changes
- can we make the search the same width as the domain requests search? I'm seeing 435.45 wide for domains and 318 for domains.
- Can we add a a margin of 12 to the right of the Export buttons
- Can we change the copy to Export as CSV
- Can we reduce the space between the export button and the text to 4?

**Comment by gabydisarli** [comment]
Accessibility question:
- shouldn't the export csv be treated as a button and not a link?

**Comment by gabydisarli** [comment]
Hey Zander!

- 383 (at max-width) and align right feels right. I want there to be more white space between the heading and the search bar, like in the domain requests table. I understand that the requests table doesn't have the export CSV button to account for, but the heading is longer so I think we can gain a little more space there.
- Yes, padding. Sorry for the confusion. Add a padding of 12 to the right of the button.
- Sorry, not "copy" (habits from my graphic design career). I mean the label of the button. Change that to "Export as CSV"

File: src/registrar/models/utility/generic_helper.py [hunk 4]
```diff
@@ -170,18 +170,11 @@ def _update_org_type_from_generic_org_and_election(self):
                 # There is no avenue for this to occur in the UI,
                 # as such - this can only occur if the object is initialized in this way.
                 # Or if there are pre-existing data.
-                logger.debug(
-                    "create_or_update_organization_type() -> is_election_board "
-                    f"cannot exist for {generic_org_type}. Setting to None."
-                )
```
[the remark this answers is not in the document]
Comment (dave-kennedy-ecs) [reply src/registrar/models/utility/generic_helper.py:176]: Hallelujah

File: src/registrar/config/urls.py [hunk 5]
```diff
@@ -9,7 +9,7 @@
 from django.views.generic import RedirectView
 
 from registrar import views
-from registrar.views.admin_views import (
+from registrar.views.report_views import (
```
Comment (dave-kennedy-ecs) [inline src/registrar/config/urls.py:12]: I think this is a very good change, which will make code more findable

File: src/registrar/utility/csv_export.py [hunk 6]
```diff
@@ -191,7 +191,7 @@ def annotate_and_retrieve_fields(
         return cls.update_queryset(queryset, **kwargs)
 
     @classmethod
-    def export_data_to_csv(cls, csv_file, start_date=None, end_date=None):
+    def export_data_to_csv(cls, csv_file, **export_kwargs):
```
[the remark this answers is not in the document]
Comment (dave-kennedy-ecs) [reply src/registrar/utility/csv_export.py:195]: This is a good refactor.  I think it is readable enough and adds some flexibility to implementing classes.

**Review by dave-kennedy-ecs — APPROVED** [review]
Deployed locally.  Tested functionality locally.  Reviewed code.  Refactors look good.  Good to go.

**Comment by gabydisarli** [comment]
This looks a lot better! Couple of small tweaks:

- decrease space between Search and Export button. I'm seeing a margin of 12px there when I inspect, see screencap
Screenshot 2024-07-25 at 2 38 14 PM
Screenshot 2024-07-25 at 2 38 24 PM

- I don't see the additional padding to the right of the export button, did you add it?

**Review by gabydisarli — APPROVED** [review]
This looks great!

**Pull request merged**